File: Classes/git/PBGitBinary.m
//
//  PBGitBinary.m
//  GitX
//
//  Created by Pieter de Bie on 04-10-08.
//  Copyright 2008 __MyCompanyName__. All rights reserved.
//

#import "PBGitBinary.h"
#import "PBEasyPipe.h"

@implementation PBGitBinary

static NSString *gitPath = nil;
static NSError *gitPathError = nil;
static NSString *cachedUserConfiguredPath = nil;

static NSString * const PBGitBinaryErrorDomain = @"PBGitBinaryErrorDomain";
static const NSInteger PBGitBinaryErrorNotFound = 1;

static dispatch_queue_t PBGitBinaryResolutionQueue(void)
{
	static dispatch_once_t onceToken;
	static dispatch_queue_t queue;
	dispatch_once(&onceToken, ^{
		queue = dispatch_queue_create("xyz.commaok.gitx.git-binary", DISPATCH_QUEUE_SERIAL);
	});
	return queue;
}

static NSString *PBGitBinarySanitizedPath(NSString *candidate)
{
	if (candidate.length == 0)
		return nil;

	NSString *trimmed = [candidate stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]];
	if (trimmed.length == 0)
		return nil;

	return [trimmed stringByStandardizingPath];
}

static NSString *PBGitBinaryValidatedExecutableAtPath(NSString *candidate)
{
	NSString *standardized = PBGitBinarySanitizedPath(candidate);
	if (standardized.length == 0)
		return nil;

	if (![[NSFileManager defaultManager] isExecutableFileAtPath:standardized])
		return nil;

	NSString *version = [PBGitBinary versionForPath:standardized];
	if (!version)
		return nil;

	return standardized;
}

static NSString *PBGitBinaryLocateGit(NSError * __autoreleasing *error)
{
	NSUserDefaults *defaults = [NSUserDefaults standardUserDefaults];
	NSString *userConfigured = [defaults stringForKey:@"gitExecutable"];
	NSString *resolved = PBGitBinaryValidatedExecutableAtPath(userConfigured);
	if (resolved)
		return resolved;

	if (userConfigured.length > 0)
		NSLog(@"PBGitBinary: user-configured git path '%@' is invalid; falling back to search paths.", userConfigured);

	char *envPath = getenv("GIT_PATH");
	if (envPath) {
		NSString *envCandidate = [NSString stringWithUTF8String:envPath];
		resolved = PBGitBinaryValidatedExecutableAtPath(envCandidate);
		if (resolved)
			return resolved;
	}

	NSString *whichPath = [PBEasyPipe outputForCommand:@"/usr/bin/which" withArgs:@[@"git"]];
	resolved = PBGitBinaryValidatedExecutableAtPath(whichPath);
	if (resolved)
		return resolved;

	for (NSString *location in [PBGitBinary searchLocations]) {
		resolved = PBGitBinaryValidatedExecutableAtPath(location);
		if (resolved)
			return resolved;
	}

	NSString *xcrunPath = [PBEasyPipe outputForCommand:@"/usr/bin/xcrun" withArgs:@[@"-f", @"git"]];
	resolved = PBGitBinaryValidatedExecutableAtPath(xcrunPath);
	if (resolved)
		return resolved;

	if (error) {
		NSDictionary *userInfo = @{ NSLocalizedDescriptionKey: @"Could not locate git executable.",
			NSLocalizedRecoverySuggestionErrorKey: [PBGitBinary notFoundError] };
		*error = [NSError errorWithDomain:PBGitBinaryErrorDomain code:PBGitBinaryErrorNotFound userInfo:userInfo];
	}

	NSLog(@"PBGitBinary: Could not find a git binary. Checked user defaults, environment, PATH, standard locations, and xcrun.");
	return nil;
}

+ (NSString *)versionForPath:(NSString *)path
{
	NSString *standardized = PBGitBinarySanitizedPath(path);
	if (standardized.length == 0)
		return nil;

	if (![[NSFileManager defaultManager] isExecutableFileAtPath:standardized])
		return nil;

	NSString *version = [PBEasyPipe outputForCommand:standardized withArgs:@[@"--version"]];
	version = [version stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]];
	if ([version hasPrefix:@"git version "])
		return [version substringFromIndex:12];

	return nil;
}

+ (NSString *)path
{
	return [self resolveGitPath:nil];
}

+ (NSString *)resolveGitPath:(NSError * __autoreleasing *)error
{
	__block NSString *resolvedPath = nil;
	__block NSError *resolvedError = nil;

	dispatch_sync(PBGitBinaryResolutionQueue(), ^{
		NSString *currentUserSetting = [[NSUserDefaults standardUserDefaults] stringForKey:@"gitExecutable"];
		if ((cachedUserConfiguredPath || currentUserSetting) && ![cachedUserConfiguredPath isEqualToString:currentUserSetting]) {
			gitPath = nil;
			gitPathError = nil;
		}
		cachedUserConfiguredPath = [currentUserSetting copy];

		if (!gitPath && !gitPathError) {
			NSError *localError = nil;
			NSString *resolved = PBGitBinaryLocateGit(&localError);
			gitPath = resolved;
			gitPathError = localError;
		}

		resolvedPath = gitPath;
		resolvedError = gitPathError;
	});

	if (!resolvedPath && error)
		*error = resolvedError;

	return resolvedPath;
}

+ (void)invalidateCachedPath
{
	dispatch_sync(PBGitBinaryResolutionQueue(), ^{
		gitPath = nil;
		gitPathError = nil;
		cachedUserConfiguredPath = nil;
	});
}

static NSMutableArray *locations = nil;

+ (NSArray *)searchLocations
{
	if (!locations)
	{
		locations = [[NSMutableArray alloc] initWithObjects:
					 @"/opt/local/bin/git",
					 @"/sw/bin/git",
					 @"/opt/git/bin/git",
					 @"/usr/local/bin/git",
					 @"/usr/local/git/bin/git",
					 @"/opt/homebrew/bin/git",
					 nil];
		
		[locations addObject:[@"~/bin/git" stringByExpandingTildeInPath]];
		[locations addObject:@"/usr/bin/git"];
	}
	return locations;
}

+ (NSString *)notFoundError
{
	NSMutableString *error = [NSMutableString stringWithString:
							  @"Could not find a git binary.\n"
							  @"Please make sure there is a git binary in one of the following locations:\n\n"];
	for (NSString *location in [PBGitBinary searchLocations]) {
		[error appendFormat:@"\t%@\n", location];
	}
	return error;
}


+ (NSString *)version
{
	NSString *path = [self resolveGitPath:nil];
	if (!path)
		return nil;

	return [self versionForPath:path];
}


@end

File: Classes/git/PBGitBinary.swift
import Foundation

/// Locates and validates the git binary.
/// Thread-safe with cached resolution that respects user preferences.
@objcMembers
@objc(PBGitBinary)
final class PBGitBinary: NSObject {

    // MARK: - Error Types

    private static let errorDomain = "PBGitBinaryErrorDomain"
    private static let errorCodeNotFound = 1

    // MARK: - Cached State

    private static let resolutionQueue = DispatchQueue(label: "xyz.commaok.gitx.git-binary")
    private static var cachedPath: String?
    private static var cachedError: NSError?
    private static var cachedUserConfiguredPath: String?

    // MARK: - Public Class Methods

    /// Returns the path to the git binary, or nil if not found.
    @objc class func path() -> String? {
        return resolveGitPath(nil)
    }

    /// Returns the git version string, or nil if git is not found.
    @objc class func gitVersion() -> String? {
        guard let gitPath = resolveGitPath(nil) else {
            return nil
        }
        return versionForPath(gitPath)
    }

    /// Returns the list of standard locations to search for git.
    @objc class func searchLocations() -> [String] {
        var locations = [
            "/opt/local/bin/git",
            "/sw/bin/git",
            "/opt/git/bin/git",
            "/usr/local/bin/git",
            "/usr/local/git/bin/git",
            "/opt/homebrew/bin/git"
        ]
        locations.append(("~/bin/git" as NSString).expandingTildeInPath)
        locations.append("/usr/bin/git")
        return locations
    }

    /// Returns an error message describing where git was searched.
    @objc class func notFoundError() -> String {
        var message = "Could not find a git binary.\n"
        message += "Please make sure there is a git binary in one of the following locations:\n\n"
        for location in searchLocations() {
            message += "\t\(location)\n"
        }
        return message
    }

    /// Resolves the git path, optionally returning an error.
    /// Thread-safe with caching that respects changes to user preferences.
    @objc class func resolveGitPath(_ error: NSErrorPointer) -> String? {
        var resolvedPath: String?
        var resolvedError: NSError?

        resolutionQueue.sync {
            let currentUserSetting = UserDefaults.standard.string(forKey: "gitExecutable")

            // Invalidate cache if user setting changed
            if (cachedUserConfiguredPath != nil || currentUserSetting != nil) &&
                cachedUserConfiguredPath != currentUserSetting {
                cachedPath = nil
                cachedError = nil
            }
            cachedUserConfiguredPath = currentUserSetting

            // Resolve if not cached
            if cachedPath == nil && cachedError == nil {
                var localError: NSError?
                let resolved = locateGit(&localError)
                cachedPath = resolved
                cachedError = localError
            }

            resolvedPath = cachedPath
            resolvedError = cachedError
        }

        if resolvedPath == nil {
            error?.pointee = resolvedError
        }

        return resolvedPath
    }

    /// Invalidates the cached path, forcing re-resolution on next access.
    @objc class func invalidateCachedPath() {
        resolutionQueue.sync {
            cachedPath = nil
            cachedError = nil
            cachedUserConfiguredPath = nil
        }
    }

    // MARK: - Private Helpers

    private class func locateGit(_ error: NSErrorPointer) -> String? {
        // 1. Check user-configured path
        if let userConfigured = UserDefaults.standard.string(forKey: "gitExecutable") {
            if let resolved = validatedExecutable(at: userConfigured) {
                return resolved
            }
            if !userConfigured.isEmpty {
                NSLog("PBGitBinary: user-configured git path '%@' is invalid; falling back to search paths.", userConfigured)
            }
        }

        // 2. Check GIT_PATH environment variable
        if let envPath = getenv("GIT_PATH") {
            let envCandidate = String(cString: envPath)
            if let resolved = validatedExecutable(at: envCandidate) {
                return resolved
            }
        }

        // 3. Use 'which git'
        if let whichPath = PBEasyPipe.outputForCommand("/usr/bin/which", withArgs: ["git"]) {
            if let resolved = validatedExecutable(at: whichPath) {
                return resolved
            }
        }

        // 4. Check standard locations
        for location in searchLocations() {
            if let resolved = validatedExecutable(at: location) {
                return resolved
            }
        }

        // 5. Try xcrun
        if let xcrunPath = PBEasyPipe.outputForCommand("/usr/bin/xcrun", withArgs: ["-f", "git"]) {
            if let resolved = validatedExecutable(at: xcrunPath) {
                return resolved
            }
        }

        // Not found
        NSLog("PBGitBinary: Could not find a git binary. Checked user defaults, environment, PATH, standard locations, and xcrun.")

        if error != nil {
            let userInfo: [String: Any] = [
                NSLocalizedDescriptionKey: "Could not locate git executable.",
                NSLocalizedRecoverySuggestionErrorKey: notFoundError()
            ]
            error?.pointee = NSError(domain: errorDomain, code: errorCodeNotFound, userInfo: userInfo)
        }

        return nil
    }

    private class func validatedExecutable(at candidate: String) -> String? {
        guard let standardized = sanitizedPath(candidate), !standardized.isEmpty else {
            return nil
        }

        guard FileManager.default.isExecutableFile(atPath: standardized) else {
            return nil
        }

        guard versionForPath(standardized) != nil else {
            return nil
        }

        return standardized
    }

    private class func sanitizedPath(_ candidate: String) -> String? {
        guard !candidate.isEmpty else {
            return nil
        }

        let trimmed = candidate.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else {
            return nil
        }

        return (trimmed as NSString).standardizingPath
    }

    private class func versionForPath(_ path: String) -> String? {
        guard let standardized = sanitizedPath(path), !standardized.isEmpty else {
            return nil
        }

        guard FileManager.default.isExecutableFile(atPath: standardized) else {
            return nil
        }

        guard var versionOutput = PBEasyPipe.outputForCommand(standardized, withArgs: ["--version"]) else {
            return nil
        }

        versionOutput = versionOutput.trimmingCharacters(in: .whitespacesAndNewlines)

        let prefix = "git version "
        if versionOutput.hasPrefix(prefix) {
            return String(versionOutput.dropFirst(prefix.count))
        }

        return nil
    }
}
